File: examples/python/client_concurrency_stages.py
#!/usr/bin/env python3
"""Concurrency stages — IdempotencyClaim, RateLimit, LockAcquire/Release.

These three primitives wrap ekoDB's atomic KV layer with the right
semantics for common server-side patterns:

    IdempotencyClaim     replay protection with TTL
    RateLimit            fixed-window counter; "fail" or "skip" mode
    LockAcquire/Release  token-fenced distributed lock

Requires ekoDB >= 0.42.0.
"""

import asyncio
import os
from pathlib import Path
from dotenv import load_dotenv

from ekodb_client import Client, Stage

env_path = Path(__file__).parent.parent / ".env"
load_dotenv(env_path)

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")


async def main() -> None:
    client = Client.new(BASE_URL, API_KEY)
    print("✓ Client created")

    # 1. Idempotent payment — claim → branch on replay → do work.
    pay = {
        "label": "conc_demo_pay",
        "name": "Idempotent payment",
        "parameters": {
            "idempotency_key": {"required": True},
            "amount": {"required": True},
        },
        "functions": [
            Stage.idempotency_claim("{{idempotency_key}}", 86400, "claim"),
            Stage.if_condition(
                {
                    "type": "FieldEquals",
                    "value": {"field": "claim.claimed", "value": False},
                },
                then_functions=[
                    Stage.return_response(
                        {
                            "status": "replay",
                            "idempotency_key": "{{idempotency_key}}",
                        },
                        200,
                    )
                ],
                else_functions=[
                    Stage.insert(
                        "charges",
                        {
                            "amount": "{{amount}}",
                            "idempotency_key": "{{idempotency_key}}",
                        },
                    ),
                    Stage.return_response({"status": "charged"}, 201),
                ],
            ),
        ],
    }
    await _save(client, pay)

    # 2. Rate-limited endpoint (fail mode).
    rl_fail = {
        "label": "conc_demo_rl_fail",
        "name": "Rate-limit (fail mode)",
        "parameters": {"user_id": {"required": True}},
        "functions": [
            Stage.rate_limit("user-{{user_id}}", 10, 60, "rl"),
            Stage.return_response({"ok": True}, 200),
        ],
    }
    await _save(client, rl_fail)

    # 3. Rate-limited endpoint (skip mode).
    rl_skip = {
        "label": "conc_demo_rl_skip",
        "name": "Rate-limit (skip mode)",
        "parameters": {"user_id": {"required": True}},
        "functions": [
            Stage.rate_limit("user-{{user_id}}", 10, 60, "rl", on_exceed="skip"),
            Stage.if_condition(
                {
                    "type": "FieldEquals",
                    "value": {"field": "rl.allowed", "value": False},
                },
                then_functions=[Stage.return_response({"status": "rate_limited"}, 429)],
                else_functions=[Stage.return_response({"status": "ok"}, 200)],
            ),
        ],
    }
    await _save(client, rl_skip)

    # 4. Distributed lock — acquire + critical section + release (token-fenced).
    lock = {
        "label": "conc_demo_lock",
        "name": "Critical section under lock",
        "parameters": {"resource": {"required": True}},
        "functions": [
            Stage.lock_acquire("{{resource}}", 30, "lock"),
            Stage.if_condition(
                {
                    "type": "FieldEquals",
                    "value": {"field": "lock.acquired", "value": False},
                },
                then_functions=[Stage.return_response({"status": "busy"}, 409)],
                else_functions=[
                    Stage.insert(
                        "lock_demo_audit", {"resource": "{{resource}}"}, ttl=60
                    ),
                    Stage.lock_release("{{resource}}", "{{lock.token}}", "release"),
                    Stage.return_response({"status": "done"}, 200),
                ],
            ),
        ],
    }
    await _save(client, lock)

    print("\nInvoke them like:")
    print(
        '  POST /api/functions/conc_demo_pay        { "idempotency_key": "...", "amount": 100 }'
    )
    print('  POST /api/functions/conc_demo_rl_fail    { "user_id": 42 }')
    print('  POST /api/functions/conc_demo_rl_skip    { "user_id": 42 }')
    print('  POST /api/functions/conc_demo_lock       { "resource": "queue:drain" }')

    for label in (
        "conc_demo_pay",
        "conc_demo_rl_fail",
        "conc_demo_rl_skip",
        "conc_demo_lock",
    ):
        try:
            await client.delete_user_function(label)
        except Exception:
            pass
    print("\n✓ Cleaned up demo functions")


async def _save(client, func):
    try:
        await client.save_user_function(func)
        print(f"✓ {func['label']} saved")
    except Exception as e:
        print(f"SaveUserFunction({func['label']}) error: {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/client_concurrency_stages.ts
/**
 * Concurrency stages — IdempotencyClaim, RateLimit, LockAcquire/Release.
 *
 * These three primitives wrap ekoDB's atomic KV layer with the right
 * semantics for common server-side patterns:
 *
 *   - IdempotencyClaim — replay protection with TTL.
 *   - RateLimit         — fixed-window counter; "fail" or "skip" mode.
 *   - LockAcquire/Release — token-fenced distributed lock.
 *
 * Requires ekoDB >= 0.42.0.
 */

import { EkoDBClient, Stage, type UserFunction } from "@ekodb/ekodb-client";
import * as dotenv from "dotenv";

dotenv.config();

const BASE_URL = process.env.API_BASE_URL || "http://localhost:8080";
const API_KEY = process.env.API_BASE_KEY || "a-test-api-key-from-ekodb";

function isNotFoundError(error: unknown): boolean {
  const message = error instanceof Error ? error.message : String(error);
  return message.includes("status 404") || /not found/i.test(message);
}

/** True when a save failed because the function label already exists (HTTP 409). */
function isAlreadyExistsError(error: unknown): boolean {
  const message = error instanceof Error ? error.message : String(error);
  return message.includes("status 409") || message.includes("already exists");
}

/** Idempotent save: create the function, or update it in place on 409. */
async function saveOrUpdateUserFunction(
  client: EkoDBClient,
  fn: UserFunction,
): Promise<void> {
  try {
    await client.saveUserFunction(fn);
  } catch (error) {
    if (!isAlreadyExistsError(error)) throw error;
    await client.updateUserFunction(fn.label, fn);
    console.log(`Function '${fn.label}' already existed — updated instead`);
  }
}

async function main() {
  const client = new EkoDBClient(BASE_URL, API_KEY);
  await client.init();
  const suffix = `${process.pid}_${Date.now()}`;
  const payLabel = `conc_demo_pay_ts_${suffix}`;
  const failLabel = `conc_demo_rl_fail_ts_${suffix}`;
  const skipLabel = `conc_demo_rl_skip_ts_${suffix}`;
  const lockLabel = `conc_demo_lock_ts_${suffix}`;
  const labels = [payLabel, failLabel, skipLabel, lockLabel];
  let primaryError: unknown;

  try {
    console.log("✓ Client created");

    // 1. Idempotent payment — claim → branch on replay → do work.
    const pay: UserFunction = {
      label: payLabel,
      name: "Idempotent payment",
      parameters: {
        idempotency_key: { required: true },
        amount: { required: true },
      },
      functions: [
        Stage.idempotencyClaim(
          `conc:pay_ts:${suffix}:{{idempotency_key}}`,
          86400,
          "claim",
        ),
        Stage.if(
          {
            type: "FieldEquals",
            value: { field: "claim.claimed", value: false },
          },
          [
            Stage.returnResponse(
              { status: "replay", idempotency_key: "{{idempotency_key}}" },
              200,
            ),
          ],
          [
            Stage.insert(`conc_charges_ts_${suffix}`, {
              amount: "{{amount}}",
              idempotency_key: "{{idempotency_key}}",
            }),
            Stage.returnResponse({ status: "charged" }, 201),
          ],
        ),
      ],
    };
    await saveOrUpdateUserFunction(client, pay);
    console.log("✓ conc_demo_pay saved");

    // 2. Rate-limited endpoint (fail mode — gate is authoritative).
    const rlFail: UserFunction = {
      label: failLabel,
      name: "Rate-limit (fail mode)",
      parameters: { user_id: { required: true } },
      functions: [
        Stage.rateLimit(`conc:rl_fail_ts:${suffix}:{{user_id}}`, 10, 60, "rl"),
        Stage.returnResponse({ ok: true }, 200),
      ],
    };
    await saveOrUpdateUserFunction(client, rlFail);
    console.log("✓ conc_demo_rl_fail saved");

    // 3. Rate-limited endpoint (skip mode — pipeline keeps running, branch on result).
    const rlSkip: UserFunction = {
      label: skipLabel,
      name: "Rate-limit (skip mode)",
      parameters: { user_id: { required: true } },
      functions: [
        Stage.rateLimit(
          `conc:rl_skip_ts:${suffix}:{{user_id}}`,
          10,
          60,
          "rl",
          "skip",
        ),
        Stage.if(
          { type: "FieldEquals", value: { field: "rl.allowed", value: false } },
          [Stage.returnResponse({ status: "rate_limited" }, 429)],
          [Stage.returnResponse({ status: "ok" }, 200)],
        ),
      ],
    };
    await saveOrUpdateUserFunction(client, rlSkip);
    console.log("✓ conc_demo_rl_skip saved");

    // 4. Distributed lock — acquire + critical section + release (token-fenced).
    const lock: UserFunction = {
      label: lockLabel,
      name: "Critical section under lock",
      parameters: { resource: { required: true } },
      functions: [
        Stage.lockAcquire(`conc:lock_ts:${suffix}:{{resource}}`, 30, "lock"),
        Stage.if(
          {
            type: "FieldEquals",
            value: { field: "lock.acquired", value: false },
          },
          [Stage.returnResponse({ status: "busy" }, 409)],
          [
            Stage.insert(
              `conc_lock_audit_ts_${suffix}`,
              { resource: "{{resource}}" },
              false,
              60,
            ),
            Stage.lockRelease(
              `conc:lock_ts:${suffix}:{{resource}}`,
              "{{lock.token}}",
              "release",
            ),
            Stage.returnResponse({ status: "done" }, 200),
          ],
        ),
      ],
    };
    await saveOrUpdateUserFunction(client, lock);
    console.log("✓ conc_demo_lock saved");

    console.log("\nInvoke them like:");
    console.log(
      `  POST /api/functions/${payLabel} { "idempotency_key": "...", "amount": 100 }`,
    );
    console.log(`  POST /api/functions/${failLabel} { "user_id": 42 }`);
    console.log(`  POST /api/functions/${skipLabel} { "user_id": 42 }`);
    console.log(
      `  POST /api/functions/${lockLabel} { "resource": "queue:drain" }`,
    );
  } catch (error) {
    primaryError = error;
    throw error;
  } finally {
    const cleanupErrors: unknown[] = [];
    for (const label of labels.reverse()) {
      try {
        await client.deleteUserFunction(label);
      } catch (error) {
        if (!isNotFoundError(error)) cleanupErrors.push(error);
      }
    }
    if (cleanupErrors.length > 0) {
      throw new AggregateError(
        primaryError === undefined
          ? cleanupErrors
          : [primaryError, ...cleanupErrors],
        "Concurrency stages example or cleanup failed",
      );
    }
    console.log("\n✓ Cleaned up demo functions");
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
